src/Hosting/TestHost/test/TestServerTests.cs

```diff
@@ -278,6 +280,58 @@ public void TestServerConstructorWithNullFeatureCollectionThrows()
             Assert.Throws<ArgumentNullException>(() => new TestServer(builder, null));
         }
 
+        [Fact]
+        public void TestServerConstructorShouldProvideServicesFromPassedServiceProvider()
+        {
+            // Arrange
+            var serviceProvider = new ServiceCollection().BuildServiceProvider();
+
+            // Act
+            var testServer = new TestServer(serviceProvider);
+
+            // Assert
+            Assert.Equal(serviceProvider, testServer.Services);
+        }
+
+        [Fact]
+        public void TestServerConstructorShouldProvideServicesFromWebHost()
+        {
+            // Arrange
+            var testService = new TestService();
+            var builder = new WebHostBuilder()
+                .ConfigureServices(services => services.AddSingleton(testService))
+                .Configure(_ => { });
+
+            // Act
+            var testServer = new TestServer(builder);
+
+            // Assert
+            Assert.Equal(testService, testServer.Services.GetService<TestService>());
+        }
+
+        [Fact]
+        public async Task TestServerConstructorShouldProvideServicesFromHostBuilder()
+        {
+            // Arrange
+            var testService = new TestService();
+            var host = await new HostBuilder()
```

Comment:
You're hitting an issue where not disposing the IHost can cause the tests to hang. We're tracking that over at  https://github.com/aspnet/Extensions/issues/1363
```suggestion
            using var host = await new HostBuilder()
```